File: ios/MineracaoMG/Services/MunicipalitiesService.swift
import Foundation

final class MunicipalitiesService {

    static let shared = MunicipalitiesService()
    private init() {}

    private var cache: [String: [String]] = [:]

    func fetch(regionId: String) async throws -> [String] {
        if let cached = cache[regionId] { return cached }

        guard let region = Region.find(regionId) else { return [] }
        let url = URL(string: "https://servicodados.ibge.gov.br/api/v1/localidades/mesorregioes/\(region.ibgeCode)/municipios")!
        let (data, _) = try await URLSession.shared.data(from: url)
        let raw = try JSONDecoder().decode([MunicipioJSON].self, from: data)
        let names = raw.map { $0.nome }.sorted { $0.localizedCompare($1) == .orderedAscending }
        cache[regionId] = names
        return names
    }

    private struct MunicipioJSON: Decodable {
        let nome: String
    }
}

File: android/app/src/main/kotlin/com/mineracaomg/MunicipalitiesService.kt
package com.mineracaomg

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URL

class MunicipalitiesService {

    private val json = Json { ignoreUnknownKeys = true }
    private val cache = mutableMapOf<String, List<String>>()

    suspend fun fetch(regionId: String): List<String> {
        cache[regionId]?.let { return it }
        val region = Region.find(regionId) ?: return emptyList()

        return withContext(Dispatchers.IO) {
            try {
                val url = "https://servicodados.ibge.gov.br/api/v1/localidades" +
                        "/mesorregioes/${region.ibgeCode}/municipios"
                val text = URL(url).openStream().bufferedReader().readText()
                val items = json.decodeFromString<List<MunicipioJson>>(text)
                val names = items.map { it.nome }.sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it })
                cache[regionId] = names
                names
            } catch (e: Exception) {
                android.util.Log.e("MunicipalitiesService", "Fetch error", e)
                emptyList()
            }
        }
    }

    @Serializable
    private data class MunicipioJson(val nome: String)
}
